### DISCUSSION TITLE
A few questions on system-level installation of MOOSE on an HPC vs Single-User installation (i.e following regular Linux installation instructions)

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#25286] [Q&A General] [answered]
I have both options, although the former will take some time to get done. I saw in the documentation that the single-user installation (i.e following the same instructions for installing on Linux) has lower performance. I was wondering if this difference is significant.

Also, I wanted to ask about the installation instructions here:

https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html

After the "module load" command, which I guess assumes that a system level installation has been done, there is no need for installing a mamba environment/ "mamba activate moose" like I typically need to do? I don't fully understand how the installation works, could you please explain what happens in each one and the differences between the two?

Also, I don't have system-level permissions, so if I need to, say "update" MOOSE (or any other operation like that) would I need to make a request each time? In short, what are the key differences between each installation type?

And I assume that after cloning the directory

```
mkdir -p ~/projects
cd ~/projects
git clone https://github.com/idaholab/moose.git
cd moose
git checkout master
```

everything else works the same (i.e creating an application, running the application, adding custom kernels/materials, etc)? 

Also, does the application folder have to be in the same "projects" folder as moose or can it be in a different folder (or even a different partition) and just refer to the moose folder?

**Comment by milljm** [comment, marked as answer]
### Specialized HPC MPI vs Conda supplied MPI

When operating on HPC Clusters, it is recommended to use that cluster's approved compilers, especially with MPI enabled binaries. Such is the case with MOOSE applications. Every cluster is differently optimized. And the administrators who built the cluster also (should) provide the compilers/MPI wrappers to properly use it.

Conda is more of a one-size-fits-all solution. It cannot be optimized for the above scenario. It just works. All the time, abet not as fast.

However! Every lab, organization, education institution employs their own crew :) Some admins cut corners, others are high-paid perfectionist. e.g. your mileage may vary when performing benchmarks between HPC supplied compilers v's Conda's. That said, I don't think anyone will be able to tell you precisely the degradation/gains using one or the other.

The more cores you use, the less you see performance scaling the way it should, when using Conda compilers. If you don't plan on solving large problems (requiring thousands of cores to solve), the less you are going to notice.

Others on my team know more about scaling... I am sure they will voice their own comments.

### HPC Install MOOSE (module load)
>After the "module load" command, which I guess assumes that a system level installation has been done, there is no need for installing a mamba environment/ "mamba activate moose" like I typically need to do?

Correct. If you are following HPC instructions, no use of Conda is involved.

The differences stem from MOOSE's requirement to be built using the same compilers used to build its dependencies (PETSc and libMesh). It is therefore the reason why you are required to build these two libraries yourself, using your HPC's compiler stack.

If you are using Conda, this is not necessary because we can supply not only PETSc and libMesh, but also the compilers used to create these two libraries.

> Also, I don't have system-level permissions, so if I need to, say "update" MOOSE (or any other operation like that) would I need to make a request each time? In short, what are the key differences between each installation type?

You should be cloning your own copy of the MOOSE repository into a place where only you have write access (your HOME directory somewhere). Meaning, it is your responsibility to update it. When you need to/want to. Some labs have tried to supply a pre-built MOOSE for their customers. In my own opinion this sounds like a nightmare.

We don't recommend this.

When developing your MOOSE based application, you are _going_ to need write access into the MOOSE repository (all those .C files during built create object files). In order for this to work in a multi-user fashion, everyone involved would need write access to the MOOSE repository. Clobbering each other previous build while _they_ work. -a nightmare.

### After cloning
> And I assume that after cloning the directory...

All correct. You can begin MOOSE based development in any directory you have write access in. `~/projects` was chosing strictly for documentation purposes. Your application does not need to co-exist next to where you cloned MOOSE. However, if it is not, you will need to set an environment variable explaining to your MOOSE based application where it is:

```bash
export MOOSE_DIR=/some/path/to/moose
# e.g. this is equivalent to the following according your demonstrated code block
export MOOSE_DIR=~/projects/moose
```

**Reply by richmondodufisan** [reply]
This is incredibly helpful, thank you!

**Reply by ABallisat** [reply]
Adding a little bit of our own experience of running stuff on shared clusters (in case it is helpful):

- If you build MOOSE with all modules in a location that others can also access, you can create a shared library that others can build against so long as they also enable all modules in their make file. They then don't need write access to that file to build so you can lock it down. This gets around the issue of everyone having to have their own local copy. It does trade off against (as far as I can tell) a small overhead but is worthwhile in my opinion. We have many users developing their apps simultaneously on the same MOOSE build without issue.
- The above also guarantees that people can use the exact same instance of MOOSE (and libmesh and PETSc) when running apps. We have had issue in the past with different people working with different local versions of MOOSE (even by only a couple of weeks) which can break things or lead to different results. This approach really eradicates that issue.
- Rolling back to older versions of MOOSE: this can be difficult as trying to match up the correct versions of build dependencies is not trivial. This approach effectively gives fixed versions that people can roll back to in future with ease. It's helped in the past with working out what changes broke some things. I know that this isn't the intended use of MOOSE but sometimes needs must when you are stuck.
- These builds are automatable, a good HPC sys admin should be able to do it, so long term it has low overhead.
- Conda was not designed for HPC, avoid it if you can. You will get significant speed up if you use the compilers and specialised built libraries your HPC sys admin supplies, don't be surprised if it is a factor of 2 or more. Also try different MPI implementations on your system, you can get very different performance depending on how they were built.
- When running on HPC nodes, you may find some performance benefit if you run slightly under max core count (at least in some cases we have). So if you have 10 cores on a node, try running with 9. I think this has something to do with leaving a core free to run the background tasks on the node and minimise context switching.

Hope that's useful!

**Reply by richmondodufisan** [reply]
It really is! Seems like a system level installation + shared folder development is the way to go. Thank you!!
